{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## shNfkb1-lane1\n",
    "\n",
    "2/2/24 - Data generated from shNfkb1 single hairpin experiments; included shRNA hairpin library, nonbarcoded; data submitted by Jason/Hannah/CHP on 5/23/2023\n",
    "\n",
    "Analyze global run on cellranger v6 data and create h5_ad files for further analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "\n",
    "import math\n",
    "import matplotlib\n",
    "\n",
    "#from solo import hashsolo\n",
    "import anndata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.settings.verbosity = 3             # verbosity: errors (0), warnings (1), info (2), hints (3)\n",
    "sc.logging.print_header()\n",
    "sc.settings.set_figure_params(dpi=80, facecolor='white')\n",
    "np.random.seed(1573)   #fix so we can reproduce later"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "results_file = 'write/2434-nfkb1.h5ad'  # the file that will store the analysis results"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Read in the count matrix into an [`AnnData`](https://anndata.readthedocs.io/en/latest/anndata.AnnData.html) object, which holds many slots for annotations and different representations of the data. It also comes with its own HDF5 file format: `.h5ad`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "adata = sc.read_10x_h5('filtered_feature_bc_matrix.h5', gex_only=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.var_names_make_unique()  # this is unnecessary if using `var_names='gene_ids'` in `sc.read_10x_mtx`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "samples = list(adata[:,adata.var['feature_types']=='Antibody Capture'].var.index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(samples)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Now filter out barcodes\n",
    "hashadata = adata[:,samples]\n",
    "adata = adata[:,[y for y in adata.var_names if y not in samples]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "hashadata.var"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Preprocessing"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Show those genes that yield the highest fraction of counts in each single cell, across all cells."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.highest_expr_genes(adata, n_top=20, )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Basic filtering:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.filter_cells(adata, min_counts=1500)\n",
    "sc.pp.filter_cells(adata, min_genes=300)\n",
    "#sc.pp.filter_genes(adata, min_cells=0.1*len(adata.obs))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's assemble some information about mitochondrial genes, which are important for quality control.\n",
    "\n",
    "Citing from \"Simple Single Cell\" workflows [(Lun, McCarthy & Marioni, 2017)](https://master.bioconductor.org/packages/release/workflows/html/simpleSingleCell.html#examining-gene-level-metrics):\n",
    "\n",
    "> High proportions are indicative of poor-quality cells (Islam et al. 2014; Ilicic et al. 2016), possibly because of loss of cytoplasmic RNA from perforated cells. The reasoning is that mitochondria are larger than individual transcript molecules and less likely to escape through tears in the cell membrane."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "With `pp.calculate_qc_metrics`, we can compute many metrics very efficiently."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.var['mt'] = adata.var_names.str.startswith('mt-') # annotate the group of mitochondrial genes as 'MT'\n",
    "sc.pp.calculate_qc_metrics(adata, qc_vars=['mt'], percent_top=None, log1p=False, inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "A violin plot of some of the computed quality measures:\n",
    "\n",
    "* the number of genes expressed in the count matrix\n",
    "* the total counts per cell\n",
    "* the percentage of counts in mitochondrial genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.violin(adata, ['n_genes_by_counts', 'total_counts', 'pct_counts_mt'],\n",
    "             jitter=0.4, multi_panel=True, rotation=90)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs[\"log10GenesPerUMI\"] = adata.obs['n_genes_by_counts'].apply(math.log10) / adata.obs['total_counts'].apply(math.log10)\n",
    "matplotlib.pyplot.hist(adata.obs[\"log10GenesPerUMI\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.violin(adata, ['log10GenesPerUMI'],\n",
    "             jitter=0.4, multi_panel=True, rotation=90)\n",
    "#99% seems like a good Human pct count cutoff"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Keep singlets, remove cells that have too many mouse reads, or mitochondrial genes expressed or too many total counts:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Filter by Log10GenesPerUMI\n",
    "adata = adata[adata.obs.log10GenesPerUMI >= 0.8,:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#adata = adata[adata.obs.n_genes_by_counts < 2500, :]\n",
    "adata = adata[adata.obs.pct_counts_mt <= 10, :]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Run Demultiplexing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "hashadata = hashadata[hashadata.obs.index.isin(adata.obs.index), :]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "hashadata.var_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "hashCounts = pd.DataFrame(hashadata.X.todense(), columns=hashadata.var_names, index=adata.obs.index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "hashDisc = hashCounts.describe([.1,.2,.3,.4,.5,.6,.7,.8,.9,0.99])\n",
    "hashDisc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "hashIDs = hashCounts.copy()\n",
    "hashID = hashadata.var_names\n",
    "for hashName in hashadata.var_names:\n",
    "    print(hashName)\n",
    "    print(hashDisc.loc[\"90%\",hashName])\n",
    "    hashIDs.loc[:,hashName] = hashCounts.loc[:,hashName] > hashDisc.loc[\"90%\",hashName]\n",
    "hashIDs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from matplotlib import pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "hashCounts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.rcParams['figure.figsize'] = (4,4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axs = plt.subplots(2,5, figsize =(20, 5))\n",
    "\n",
    "for i, hashName in enumerate(hashadata.var_names):\n",
    "    hashCounts2 = np.log10(hashCounts[hashName]+1)\n",
    "    axs[i//5,i%5].set_title(hashName)\n",
    "    axs[i//5,i%5].hist(hashCounts2, bins = 100)\n",
    "    axs[i//5,i%5].axvline(np.log10(int(hashDisc.loc[\"90%\",hashName])+1), color='k', linestyle='dashed', linewidth=1)\n",
    "\n",
    "#plt.rcParams[\"figure.figsize\"] = (20,5)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "hashName= \"AW1781_Renilla\"\n",
    "hash301 = np.log10(hashCounts[hashName]+1)\n",
    "\n",
    "fig, ax = plt.subplots(figsize =(10, 7))\n",
    "ax.hist(hash301, bins = 100)\n",
    "ax.axvline(np.log10(int(hashDisc.loc[\"90%\",hashName])+1), color='k', linestyle='dashed', linewidth=1)\n",
    "# Show plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.mixture import BayesianGaussianMixture\n",
    "from scipy.stats import norm\n",
    "fig, axs = plt.subplots(2,5, figsize =(20, 5))\n",
    "dfHashBoundry = pd.DataFrame(np.zeros(len(hashadata.var_names)),hashadata.var_names, columns=[\"boundry\"])\n",
    "gmm = BayesianGaussianMixture(n_components=2, random_state=250,init_params='k-means++')\n",
    "binEx = np.arange(0,5,5/100).reshape(-1,1)\n",
    "\n",
    "for i, hashName in enumerate(hashadata.var_names):\n",
    "    hashCount = np.array(np.log10(hashCounts[hashName]+1)).reshape(-1, 1)\n",
    "    fitGMM = gmm.fit(hashCount)\n",
    "    mean = fitGMM.means_  \n",
    "    covs  = fitGMM.covariances_\n",
    "    weights = fitGMM.weights_\n",
    "  \n",
    "    fitGmmBound = fitGMM.predict(binEx)\n",
    "\n",
    "    x_axis = np.arange(0, 5, 0.1)\n",
    "    y_axis0 = norm.pdf(x_axis, float(mean[0][0]), np.sqrt(float(covs[0][0][0])))*weights[0] # 1st gaussian\n",
    "    y_axis1 = norm.pdf(x_axis, float(mean[1][0]), np.sqrt(float(covs[1][0][0])))*weights[1] # 2nd gaussian\n",
    "\n",
    "    hashBoundry = False #binEx[np.where(fitGmmBound == 1)[0][0]][0]    \n",
    "    if mean[0][0] < mean[1][0]:\n",
    "        hashBoundry = x_axis[np.where(y_axis1 < y_axis0)[0][-1]+2]\n",
    "    else:\n",
    "        hashBoundry = x_axis[np.where(y_axis0 < y_axis1)[0][-1]+2]\n",
    "    \n",
    "    naiveBoundry = np.log10(int(hashDisc.loc[\"90%\",hashName])+1)\n",
    "    \n",
    "    dfHashBoundry.loc[hashName] = hashBoundry\n",
    "    \n",
    "    # Plot 2\n",
    "    axs[i//5,i%5].set_title(hashName)\n",
    "    axs[i//5,i%5].axvline(naiveBoundry, c='C3', linestyle='dashed', linewidth=1) #red\n",
    "    axs[i//5,i%5].axvline(hashBoundry, c='C2', linestyle='dashed', linewidth=1)  #green\n",
    "    axs[i//5,i%5].hist(hashCount, density=True, color='black', bins=100)        \n",
    "    axs[i//5,i%5].plot(x_axis, y_axis0, lw=3, c='C6')                            #pink\n",
    "    axs[i//5,i%5].plot(x_axis, y_axis1, lw=3, c='C1')                            #orange\n",
    "    axs[i//5,i%5].plot(x_axis, y_axis0+y_axis1, lw=3, c='C0', ls=':')            #dotted blue\n",
    "    \n",
    "plt.tight_layout(pad=1.0)\n",
    "#plt.rcParams[\"figure.figsize\"] = (20,5)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "hashIDs = hashCounts.copy()\n",
    "hashID = hashadata.var_names\n",
    "for hashName in hashadata.var_names:\n",
    "    print(hashName)\n",
    "    print(dfHashBoundry.loc[hashName].values[0])\n",
    "    hashIDs.loc[:,hashName] = np.log10(hashCounts.loc[:,hashName]+1) > dfHashBoundry.loc[hashName].values[0]\n",
    "#hashIDs.loc[:,'AT1129_B0304_2117'] = False\n",
    "#hashIDs.loc[:,'AT1606_B0305_Renilla'] = False\n",
    "#hashIDs.loc[:,'AT1831_B0306_2117'] = False\n",
    "#hashIDs.loc[:,'AT1607_B0307_Renilla'] = False\n",
    "#hashIDs.loc[:,'AV1189_B0310_2118'] = False\n",
    "hashIDs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "classification = np.empty(len(adata), dtype=\"object\")\n",
    "i = 0\n",
    "for cellBar, hashBool in hashIDs.iterrows():\n",
    "    #print(i)\n",
    "    #print(hashBool)\n",
    "    #print(hashBool.values)\n",
    "    #print(sum(hashBool))\n",
    "    numHashes = sum(hashBool)\n",
    "    if (numHashes == 1):\n",
    "        classif = hashID[hashBool.values].values[0]\n",
    "    elif (numHashes > 1):\n",
    "        classif = \"doublet\"\n",
    "    else:\n",
    "        classif = \"negative\"\n",
    "    classification[i] = classif\n",
    "    i = i + 1\n",
    "    #break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "adata.obs[\"Classification\"] = classification\n",
    "hashadata.obs[\"Classification\"] = classification\n",
    "adata.obs[\"Classification\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#output visulaization of hashing\n",
    "sc.pl.heatmap(hashadata, hashadata.var_names, groupby=\"Classification\", log=True)#, save = f\"_{figName}_hash.png\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Scrublet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "import scrublet as scr\n",
    "import scipy.io"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "doublet_calls = np.array(adata.obs.Classification) == 'doublet'\n",
    "naive_doublet_rate = doublet_calls.sum() / float(len(doublet_calls))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "expected_doublet_rate = naive_doublet_rate * len(samples) / (len(samples)-1)\n",
    "print(\"Expected doublet rate: %0.2f\" % expected_doublet_rate)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "scrub = scr.Scrublet(adata.X, expected_doublet_rate=0.128)\n",
    "#scrub = scr.Scrublet(adata.X, expected_doublet_rate=expected_doublet_rate)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "doublet_scores, predicted_doublets = scrub.scrub_doublets(n_prin_comps=30, \n",
    "                                                            mean_center=True, \n",
    "                                                            normalize_variance=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#scrub.call_doublets(threshold=0.3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "scrub.plot_histogram();"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print('Running UMAP...')\n",
    "scrub.set_embedding('UMAP', scr.get_umap(scrub.manifold_obs_, 10, min_dist=0.3))\n",
    "\n",
    "# # Uncomment to run tSNE - slow\n",
    "# print('Running tSNE...')\n",
    "# scrub.set_embedding('tSNE', scr.get_tsne(scrub.manifold_obs_, angle=0.9))\n",
    "\n",
    "# # Uncomment to run force layout - slow\n",
    "# print('Running ForceAtlas2...')\n",
    "# scrub.set_embedding('FA', scr.get_force_layout(scrub.manifold_obs_, n_neighbors=5. n_iter=1000))\n",
    "    \n",
    "print('Done.')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "scrub.plot_embedding('UMAP', order_points=True);\n",
    "\n",
    "# scrub.plot_embedding('tSNE', order_points=True);\n",
    "# scrub.plot_embedding('FA', order_points=True);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "doub_calls = {'Doublet': doublet_calls,\n",
    "              'Scrublet (z-score)': scrub.predicted_doublets_, \n",
    "              }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "x = scrub._embeddings['UMAP'][:,0]\n",
    "y = scrub._embeddings['UMAP'][:,1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axs = plt.subplots(1, 2, figsize = (14, 5))\n",
    "for iMethod, method in enumerate(doub_calls):\n",
    "    coldat = doub_calls[method]\n",
    "    o = np.argsort(coldat)\n",
    "\n",
    "    ax = axs[iMethod]\n",
    "    ax.scatter(x[o], y[o], c = coldat[o], cmap=scr.custom_cmap([[.7,.7,.7], [0,0,0]]), s = 2)\n",
    "    ax.set_xticks([]), ax.set_yticks([])\n",
    "    ax.set_title(method)\n",
    "fig.tight_layout()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs['Scrublet_doublet']=scrub.predicted_doublets_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    " cell_proportion_df = pd.crosstab(adata.obs['Classification'],adata.obs['Scrublet_doublet'], normalize='columns').T.plot(kind='bar', stacked=True).legend(bbox_to_anchor=(1,1))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Examine raw data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "hadata = adata.copy()"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "sc.pp.filter_cells(hadata, min_genes=200)\n",
    "sc.pp.filter_genes(hadata, min_cells=3)\n",
    "sc.pp.filter_cells(hadata, min_counts=1500)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "sc.pp.normalize_total(hadata, target_sum=1e4)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Logarithmize the data:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "sc.pp.log1p(hadata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "sc.pp.highly_variable_genes(hadata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "hadata = hadata[:, hadata.var.highly_variable]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "sc.pp.scale(hadata, max_value=10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Principal component analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Reduce the dimensionality of the data by running principal component analysis (PCA), which reveals the main axes of variation and denoises the data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.tl.pca(hadata, svd_solver='arpack', n_comps=150)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can make a scatter plot in the PCA coordinates, but we will not use that later on."
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "sc.pl.pca(hadata)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let us inspect the contribution of single PCs to the total variance in the data. This gives us information about how many PCs we should consider in order to compute the neighborhood relations of cells, e.g. used in the clustering function  `sc.tl.louvain()` or tSNE `sc.tl.tsne()`. In our experience, often a rough estimate of the number of PCs does fine."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.pca_variance_ratio(hadata, log=True, n_pcs=150)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Computing the neighborhood graph"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let us compute the neighborhood graph of cells using the PCA representation of the data matrix. You might simply use default values here. For the sake of reproducing Seurat's results, let's take the following values."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.neighbors(hadata, n_neighbors=15, n_pcs=75)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Embedding the neighborhood graph"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We suggest embedding the graph in two dimensions using UMAP ([McInnes et al., 2018](https://arxiv.org/abs/1802.03426)), see below. It is potentially more faithful to the global connectivity of the manifold than tSNE, i.e., it better preserves trajectories. In some ocassions, you might still observe disconnected clusters and similar connectivity violations. They can usually be remedied by running:\n",
    "\n",
    "```\n",
    "tl.paga(adata)\n",
    "pl.paga(adata, plot=False)  # remove `plot=False` if you want to see the coarse-grained graph\n",
    "tl.umap(adata, init_pos='paga')\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.tl.umap(hadata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "sc.pl.umap(hadata, color='Classification', groups=['doublet','negative'])\n",
    "#for label in hadata.obs['Classification'].unique():\n",
    "#    sc.pl.umap(hadata, color='Classification',groups=[label])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "sc.pl.umap(hadata, color='Classification')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(hadata, color='Classification', groups=samples)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "sc.pl.violin(hadata,['2117-Nfkb1.2331','2118-Nfkb1.3737'], groupby='Classification', rotation=90)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "hadata.obs.Classification.value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## set highly variable genes before moving all mean to 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#Remove any cell with PTPRC counts\n",
    "fig, ax = plt.subplots(figsize =(10, 7))\n",
    "ax.hist(adata.X[:,adata.var.index.isin(['shRenilla'])].toarray())\n",
    "#ax.axvline(np.log10(int(hashDisc.loc[\"90%\",hashName])+1), color='k', linestyle='dashed', linewidth=1)\n",
    "# Show plot\n",
    "#adata.obs['shRen'] = 0\n",
    "\n",
    "#adata.obs['shRenilla'][adata.X[:,adata.var.index.isin(['shRenilla'])] >0] #= adata.X[:,adata.var.index.isin(['shRenilla'])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs['shRen'] = adata.X[:,adata.var.index.isin(['shRenilla'])].toarray()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#Remove any cell with PTPRC counts\n",
    "fig, ax = plt.subplots(figsize =(10, 7))\n",
    "ax.hist(adata.X[:,adata.var.index.isin(['2118-Nfkb1.3737'])].toarray())\n",
    "#ax.axvline(np.log10(int(hashDisc.loc[\"90%\",hashName])+1), color='k', linestyle='dashed', linewidth=1)\n",
    "# Show plot\n",
    "#adata.obs['2118-Nfkb1.3737'] = 0\n",
    "\n",
    "#adata.obs['shRenilla'][adata.X[:,adata.var.index.isin(['shRenilla'])] >0] #= adata.X[:,adata.var.index.isin(['shRenilla'])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs['2118-Nfkb1'] = adata.X[:,adata.var.index.isin(['2118-Nfkb1.3737'])].toarray()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#Remove any cell with PTPRC counts\n",
    "fig, ax = plt.subplots(figsize =(10, 7))\n",
    "ax.hist(adata.X[:,adata.var.index.isin(['2117-Nfkb1.2331'])].toarray())\n",
    "#ax.axvline(np.log10(int(hashDisc.loc[\"90%\",hashName])+1), color='k', linestyle='dashed', linewidth=1)\n",
    "# Show plot\n",
    "#adata.obs['2117-Nfkb1.2331'] = 0\n",
    "\n",
    "#adata.obs['shRenilla'][adata.X[:,adata.var.index.isin(['shRenilla'])] >0] #= adata.X[:,adata.var.index.isin(['shRenilla'])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs['2117-Nfkb1'] = adata.X[:,adata.var.index.isin(['2117-Nfkb1.2331'])].toarray()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "adata.var.index.isin(['GFP'])\n",
    "_ = plt.hist(adata[adata.obs.index,'GFP'].X.toarray())\n",
    "plt.ylabel(\"Count\")\n",
    "plt.title(\"GFP\")\n",
    "plt.axvline(x=2)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs['GFP_count'] = adata.X[:,adata.var.index.isin(['GFP'])].toarray()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "adata.var.index.isin(['GFP'])\n",
    "_ = plt.hist(adata[adata.obs.index,'Cre'].X.toarray())\n",
    "plt.ylabel(\"Count\")\n",
    "plt.title(\"Cre\")\n",
    "plt.axvline(x=2)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs['Cre_count'] = adata.X[:,adata.var.index.isin(['Cre'])].toarray()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Filter doublets and negative cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "adata.obs.Classification.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = adata[~(adata.obs['Classification'].isin(['doublet','negative']) | adata.obs.Scrublet_doublet),:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs.Classification.value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Filter genes ( >= 1% of cells )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "sc.pp.filter_genes(adata, min_cells=0.01*len(adata.obs))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Set Groups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Set shRNA groups\n",
    "adata.obs['shRNA'] = None\n",
    "adata.obs['shRNA'][adata[adata.obs['shRen'] >= 2,:].obs.index] = 'shRenilla'\n",
    "adata.obs['shRNA'][adata[adata.obs['2117-Nfkb1'] >= 2,:].obs.index] = '2117'\n",
    "adata.obs['shRNA'][adata[adata.obs['2118-Nfkb1'] >= 2,:].obs.index] = '2118'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Set shRNA groups\n",
    "adata.obs['Group'] = 'Renilla_Control'\n",
    "adata.obs['Group'][adata.obs.Classification.isin(['AW1781_Renilla','AX1068_Renilla'])] = 'Renilla_EGFRi'\n",
    "adata.obs['Group'][adata.obs.Classification.isin(['AX1127_2117','AX1893_2117'])] = '2117_EGFRi'\n",
    "adata.obs['Group'][adata.obs.Classification.isin(['AW1785_2118'])] = '2118_EGFRi'\n",
    "adata.obs['Group'][adata.obs.Classification.isin(['AX1508_2117','AX1598_2117'])] = '2117_Control'\n",
    "adata.obs['Group'][adata.obs.Classification.isin(['AW1555_2118'])] = '2118_Control'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Double check shRNAs and filter accordingly"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "sc.pl.violin(adata,['shRen','2117-Nfkb1','2118-Nfkb1'], groupby='Classification', rotation=90)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs[['shRNA','Group']].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = adata[~((adata.obs.shRNA.isin(['2117','shRenilla']) & adata.obs.Group.isin(['2118_EGFRi'])) | \n",
    "                (adata.obs.shRNA.isin(['shRenilla']) & adata.obs.Group.isin(['2117_EGFRi','2117_Control']))),:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs[['shRNA','Group']].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.violin(adata,['shRen','2117-Nfkb1','2118-Nfkb1'], groupby='Group', rotation=90)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.violin(adata, ['n_genes_by_counts', 'total_counts', 'pct_counts_mt'],\n",
    "             jitter=0.4, multi_panel=True, rotation=90)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.layers['counts']=adata.X.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "sc.pp.normalize_total(adata, target_sum=1e4)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Logarithmize the data:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "sc.pp.log1p(adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "sc.pp.highly_variable_genes(adata) #batch_key=Classification"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "sc.pl.highly_variable_genes(adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library = [\"1659-Fosl2.226\",\"1660-Fosl2.752\",\"1661-Fhl2.467\",\"1662-Fhl2.179\",\"1663-Myc.989\",\"1664-Myc.1389\",\"1665-Yap1.735\",\"1666-Yap1.2847\",\"1667-Atf4.1383\",\"1668-Atf4.536\",\"1669-Maff.85\",\"1670-Maff.587\",\"1671-Nfkb2.152\",\"1672-Nfkb2.534\",\"1673-Relb.286\",\"1674-Relb.1795\",\"1675-Onecut2.1799\",\"1676-Onecut2.2361\",\"1810-Slc4a11.541\",\"1934-Slc4a11.1966\",\"1811-Itga2.672\",\"1812-Itga2.204\",\"1998-Cldn4-1013\",\"1999-Cldn4-951\",\"2113-Lif.440\",\"2114-Lif.1890\",\"2115-Fosl1.1401\",\"2116-Fosl1.699\",\"2117-Nfkb1.2331\",\"2118-Nfkb1.3737\",\"2190-Jun_2481\",\"2191-Jun_3136\",\"2193-JunB_406\",\"2194-JunD_78\",\"2195-JunD_859\",\"JCP618-Rela_1202\",\"JCP621-Rela_1182\",\"JCP623-Rel_1868\",\"JCP625-Rel_342\",\"JCP628-Ikbkg_2526\",\"JCP629-Ikbkg_2984\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#Ensure marker genes aren't making an impact to highly variable gene analysis downstream\n",
    "adata.var.highly_variable['GFP'] = False\n",
    "adata.var.highly_variable['mKate2'] = False\n",
    "adata.var.highly_variable['Cre'] = False\n",
    "for x in library:\n",
    "    adata.var.highly_variable[x] = False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.raw = adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.write('write/allmice.h5ad', compression='gzip')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = sc.read('write/allmice.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = adata[:, adata.var.highly_variable]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "sc.pp.regress_out(adata, ['total_counts', 'pct_counts_mt'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "sc.pp.scale(adata, max_value=10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Principal component analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Reduce the dimensionality of the data by running principal component analysis (PCA), which reveals the main axes of variation and denoises the data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.tl.pca(adata, svd_solver='arpack', n_comps=150)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can make a scatter plot in the PCA coordinates, but we will not use that later on."
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {
    "scrolled": true
   },
   "source": [
    "sc.pl.pca(adata, color='Cst3')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let us inspect the contribution of single PCs to the total variance in the data. This gives us information about how many PCs we should consider in order to compute the neighborhood relations of cells, e.g. used in the clustering function  `sc.tl.louvain()` or tSNE `sc.tl.tsne()`. In our experience, often a rough estimate of the number of PCs does fine."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.pca_variance_ratio(adata, log=True, n_pcs=150)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Save the result."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.write(results_file, compression='gzip')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Computing the neighborhood graph"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let us compute the neighborhood graph of cells using the PCA representation of the data matrix. You might simply use default values here. For the sake of reproducing Seurat's results, let's take the following values."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "sc.pp.neighbors(adata, n_neighbors=15, n_pcs=70)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Embedding the neighborhood graph"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We suggest embedding the graph in two dimensions using UMAP ([McInnes et al., 2018](https://arxiv.org/abs/1802.03426)), see below. It is potentially more faithful to the global connectivity of the manifold than tSNE, i.e., it better preserves trajectories. In some ocassions, you might still observe disconnected clusters and similar connectivity violations. They can usually be remedied by running:\n",
    "\n",
    "```\n",
    "tl.paga(adata)\n",
    "pl.paga(adata, plot=False)  # remove `plot=False` if you want to see the coarse-grained graph\n",
    "tl.umap(adata, init_pos='paga')\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.tl.umap(adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata, color=['Classification'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "sc.pl.umap(adata, color=['Epcam','Ptprc','Vim', 'Cdh1','Cdh2']) #'Ptprc',"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata, color=['Kras','Trp53'], vmax=5)\n",
    "sc.pl.umap(adata, color=['mKate2','GFP'], vmax=2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata, color=['Hopx','Pdpn'], vmax=5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata, color=['total_counts'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Score Cluster 5 cells\n",
    "import csv\n",
    "clusters = {}\n",
    "clusterkeys = []\n",
    "HPCS = False\n",
    "\n",
    "def resetClusters(hpcs = 'cell2020'):\n",
    "    clusters = {}\n",
    "    clusterkeys = []\n",
    "    HPCS = False\n",
    "    try:\n",
    "        if hpcs == 'tumorClusters':\n",
    "            with open('clusters/tumorClustersv2.csv',encoding='utf-8-sig') as csvfile:\n",
    "                csvreader = csv.reader(csvfile, delimiter=\",\")\n",
    "                for row in csvreader:\n",
    "                    clusters[row[0]] = row[1:]\n",
    "            for i in range(0,16):\n",
    "                clusterkeys.append('%i' % i)\n",
    "            HPCS = '3'\n",
    "        elif hpcs == 'cell2020':         \n",
    "            with open('clusters/clusters_cell2020.csv',encoding='utf-8-sig') as csvfile:\n",
    "                csvreader = csv.reader(csvfile, delimiter=\",\")\n",
    "                for row in csvreader:\n",
    "                    clusters[row[0]] = [x for x in row[1:] if x != '']\n",
    "            for i in range(1,13):\n",
    "                #if i == 9: continue\n",
    "                clusterkeys.append('Cluster %i' % i)\n",
    "            HPCS = 'Cluster 5'\n",
    "        elif hpcs == 'cl5bootstrap':\n",
    "            with open('clusters/cl5bootstrap.csv',encoding='utf-8-sig') as csvfile:\n",
    "                csvreader = csv.reader(csvfile, delimiter=\",\")\n",
    "                for row in csvreader:\n",
    "                    clusters[row[0]] = row[1:]\n",
    "            #for i in range(0,1):\n",
    "                #clusterkeys.append('%i' % i)\n",
    "            clusterkeys.append('Cl5Bootstrap')\n",
    "            HPCS = 'Cl5Bootstrap'\n",
    "        elif hpcs == 'cl5bootstrap2':\n",
    "            with open('clusters/cl5bootstrap2.csv',encoding='utf-8-sig') as csvfile:\n",
    "                csvreader = csv.reader(csvfile, delimiter=\",\")\n",
    "                for row in csvreader:\n",
    "                    clusters[row[0]] = row[1:]\n",
    "            #for i in range(0,1):\n",
    "                #clusterkeys.append('%i' % i)\n",
    "            clusterkeys.append('Cl5Bootstrap2')\n",
    "            HPCS = 'Cl5Bootstrap2'\n",
    "        elif hpcs == 'shRenilla':\n",
    "            raise ValueError\n",
    "            ###\n",
    "        else:\n",
    "            raise ValueError\n",
    "    except ValueError:\n",
    "        print(\"%s is an invalid choice\" % hpcs)\n",
    "        raise\n",
    "    return (clusters, clusterkeys, HPCS)\n",
    "\n",
    "(clusters, clusterkeys, HPCS) = resetClusters('cell2020')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def scoreAndPlot(ad, excludeList = None, groupby=\"Classification\",rotation=90,numgenes=25,ctlgenes=25):\n",
    "    #cmap = 'Reds' #colormap\n",
    "    cmap = 'jet' #colormap\n",
    "    if excludeList == None:\n",
    "        for i in clusterkeys:\n",
    "            if (numgenes > ctlgenes):\n",
    "                ctlgenes = numgenes\n",
    "            sc.tl.score_genes(ad, clusters[i][0:numgenes],score_name=\"%s\" % i, ctrl_size=ctlgenes)  \n",
    "        #sc.tl.score_genes(ad, clusters['Highly_mixed'],score_name=\"Highly_mixed\")\n",
    "  \n",
    "        sc.pl.umap(ad, color=clusterkeys, color_map=cmap)\n",
    "        #sc.pl.violin(ad, ['SLC4A11','ITGA2','CLDN4','TGIF1','Cluster5'],\n",
    "         #   jitter=0.4, multi_panel=True, groupby=groupby, rotation=rotation)     \n",
    "        #sc.pl.violin(ad, ['Highly_mixed'],\n",
    "        #    jitter=0.4, multi_panel=True, groupby=groupby, rotation=rotation)            \n",
    "        #sc.pl.violin(ad, ['MYC_sig','MYC','Fredlund','Jung'],\n",
    "        #    jitter=0.4, multi_panel=True, groupby=groupby, rotation=rotation)     \n",
    "        #sc.pl.violin(ad, ['ONECUT2','ONECUT2_369_sig','ONECUT2_29_sig'],\n",
    "        #    jitter=0.4, multi_panel=True, groupby=groupby, rotation=rotation) \n",
    "        #sc.pl.violin(ad, ['YAP1', 'TAZ', 'YAP_sig'],\n",
    "        #    jitter=0.4, multi_panel=True, groupby=groupby, rotation=rotation) \n",
    "        sc.pl.dotplot(ad, clusterkeys, groupby=groupby, swap_axes=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.stats import ranksums\n",
    "\n",
    "def HPCSViolinPlot(ad, cluster='0', groupby='leiden', score='Cluster5', save=None, singleGene=False):\n",
    "    ad.obs['Cl5'] = 'not HPCS'\n",
    "    ad.obs.loc[ad.obs[groupby].isin([cluster]), 'Cl5'] = 'HPCS'\n",
    "\n",
    "    if singleGene:\n",
    "        pvalue = ranksums(ad[ad.obs['Cl5'].isin(['HPCS'])][:,score].X.toarray(),ad[~ad.obs['Cl5'].isin(['HPCS'])][:,score].X.toarray())[1]\n",
    "        #sc.pl.violin(ad, score, groupby='Cl5', xlabel = 'p = ' + str(pvalue), save=save)\n",
    "    else:\n",
    "        pvalue = ranksums(ad[ad.obs['Cl5'].isin(['HPCS'])].obs[score],ad[ad.obs['Cl5'].isin(['not HPCS'])].obs[score])[1]\n",
    "    sc.pl.violin(ad, score, groupby='Cl5', xlabel = 'p = ' + str(pvalue), save=save)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "scoreAndPlot(adata, numgenes=500)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "sc.pl.umap(adata,color='Cluster 5', cmap='Reds')"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "sc.pl.umap(adata,color='Cluster 11', cmap='Reds')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Clustering the neighborhood graph"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As with Seurat and many other frameworks, we recommend the Leiden graph-clustering method (community detection based on optimizing modularity) by [Traag *et al.* (2018)](https://scanpy.readthedocs.io/en/latest/references.html#traag18). Note that Leiden clustering directly clusters the neighborhood graph of cells, which we already computed in the previous section."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.tl.leiden(adata, resolution=0.3)\n",
    "#sc.tl.leiden(adata, resolution=0.25)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Plot the clusters, which agree quite well with the result of Seurat."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata, color=['leiden', 'Epcam', 'Sftpc', # Tumor cells\n",
    "                         'Ptprc']), #immune\n",
    "                         #  'Cd4', 'Cd8a', #T cell #'Cd3e',\n",
    "                         # 'Cd19', 'Ms4a1', 'Cd22', # B cell\n",
    "                         # #'Itgax', 'Il3ra', # Dendritic Cell\n",
    "                         # 'Ncam1',# 'Ncr1', # NK Cell\n",
    "                         # 'Cd34', # Hematopoetic stem cell\n",
    "                         # #'Cd14', 'Cd33', # macrophage\n",
    "                         # #'Ceacam8' # Granulocyte\n",
    "                         # 'Itga2b', 'Itgb3', 'Selp', #platelet\n",
    "                         # #'Gypa',  # erythrocyte\n",
    "                         # #'Mcam', 'Vcam1', 'Pecam1', 'Sele', # endothelial cell\n",
    "                         # 'Cd109', 'Wnt5a', 'Kras'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "sc.pl.umap(adata, color=['Kras',  'Zeb2', ]) #'Cd109', #'Sox11'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Save the result."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata, color=['Slc4a11', 'Itga2', 'Cldn4','Epcam'], vmax=2) #'Tigit', "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata, color=['Vim', 'Epcam', 'Krt8', 'Krt18', 'Pdpn', 'Hopx'], vmax=2) #'Pecam1',"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {
    "scrolled": true
   },
   "source": [
    "sc.pl.umap(adata, color=['leiden'], groups=['6'])"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "PtprcCells = adata[adata.obs.leiden.isin(['6']),:].obs.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata, color=['GFP'], cmap='Reds')\n",
    "sc.pl.umap(adata, color=['mKate2'], cmap='Reds')"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "sc.pl.umap(adata, color=['Slc4a11'], cmap='Reds')"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "sc.pl.umap(adata, color=['Nfkb1'], cmap='Reds')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata, color=[\"shRenilla\", \"2118-Nfkb1\", \"2117-Nfkb1\"], cmap='Reds')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "for label in adata.obs['Classification'].unique():\n",
    "    sc.pl.umap(adata, color='Classification',groups=[label])"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "#fix a bug; 'base' is None but that doesn't seem to have transferred\n",
    "adata.uns['log1p'] = {}\n",
    "adata.uns['log1p']['base'] = None"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "sc.tl.rank_genes_groups(adata, 'leiden', method='t-test')\n",
    "sc.pl.rank_genes_groups(adata, n_genes=25, sharey=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Cluster 1 appears to be fibroblasts based on Krt8 and Krt18 weak staining"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "sc.pl.umap(adata,color=['n_genes', 'n_genes_by_counts', 'total_counts', 'total_counts_mt',\n",
    "       'pct_counts_mt', 'log10GenesPerUMI'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata, color=['leiden','mKate2'],cmap=\"Reds\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "sc.pl.umap(adata, color=['leiden'], groups=['0','1','2','5','6','7','10'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## End Analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Select only those subset of cells that are useful\n",
    "adata2 = sc.read('write/allmice.h5ad')\n",
    "adata = adata2[adata.obs['leiden'].isin(['0','1','2','5','6','7','10']),:] # keep tumor cells\n",
    "del adata2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#fix a bug; 'base' is None but that doesn't seem to have transferred\n",
    "adata.uns['log1p'] = {}\n",
    "adata.uns['log1p']['base'] = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.highly_variable_genes(adata)#, batch_key='Classification')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#Ensure marker genes aren't making an impact to highly variable gene analysis downstream\n",
    "adata.var.highly_variable['GFP'] = False\n",
    "adata.var.highly_variable['mKate2'] = False\n",
    "adata.var.highly_variable['Cre'] = False\n",
    "for x in library:\n",
    "    adata.var.highly_variable[x] = False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.highly_variable_genes(adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.raw = adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.write('write/allmice-tumor.h5ad', compression='gzip')\n",
    "adata = sc.read('write/allmice-tumor.h5ad')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Now rerun algorithm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = adata[:, adata.var.highly_variable]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "sc.pp.regress_out(adata, ['total_counts', 'pct_counts_mt'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "sc.pp.scale(adata, max_value=10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Principal component analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Reduce the dimensionality of the data by running principal component analysis (PCA), which reveals the main axes of variation and denoises the data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.tl.pca(adata, svd_solver='arpack', n_comps=150)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can make a scatter plot in the PCA coordinates, but we will not use that later on."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.pca(adata, color='Cst3')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let us inspect the contribution of single PCs to the total variance in the data. This gives us information about how many PCs we should consider in order to compute the neighborhood relations of cells, e.g. used in the clustering function  `sc.tl.louvain()` or tSNE `sc.tl.tsne()`. In our experience, often a rough estimate of the number of PCs does fine."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.pca_variance_ratio(adata, log=True, n_pcs=150)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Save the result."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.write(results_file, compression='gzip')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Computing the neighborhood graph"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let us compute the neighborhood graph of cells using the PCA representation of the data matrix. You might simply use default values here. For the sake of reproducing Seurat's results, let's take the following values."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.neighbors(adata, n_neighbors=15, n_pcs=50)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Embedding the neighborhood graph"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We suggest embedding the graph in two dimensions using UMAP ([McInnes et al., 2018](https://arxiv.org/abs/1802.03426)), see below. It is potentially more faithful to the global connectivity of the manifold than tSNE, i.e., it better preserves trajectories. In some ocassions, you might still observe disconnected clusters and similar connectivity violations. They can usually be remedied by running:\n",
    "\n",
    "```\n",
    "tl.paga(adata)\n",
    "pl.paga(adata, plot=False)  # remove `plot=False` if you want to see the coarse-grained graph\n",
    "tl.umap(adata, init_pos='paga')\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.tl.umap(adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "sc.pl.umap(adata, color=['Classification'])\n",
    "#sc.pl.umap(adata, color=['Porcn', 'Lgr5', 'Lgr6','Axin2']) #LGR5 not found #\n",
    "sc.pl.umap(adata, color=['Epcam', 'Ptprc', 'Vim', 'Cdh1','Cdh2']) #CDH2 not found\n",
    "sc.pl.umap(adata, color=['Slc4a11', 'Itga2', 'Cldn4']) #, 'Abcb1'\n",
    "sc.pl.umap(adata, color=['Hopx', 'Pdpn'])\n",
    "#sc.pl.umap(adata, color=['Cd81', 'Cd19'])\n",
    "#sc.pl.umap(adata, color=['Trp53','Meg3'], vmax=5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata, color=['Vim', 'Onecut2', 'mKate2', 'GFP']) #'Sox11', "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#NFkB targets\n",
    "nfkb = ['Tnf', 'Mmp13','Nfkbia']\n",
    "nfkball = ['Tnf', 'Mmp13', 'Nfkbia', 'Bcl2', 'Bcl2l1', 'Xiap', 'Birc2', 'Il6', 'Birc5', 'Ccnd1', 'Mmp3']\n",
    "nfkbfactors = ['Nfkb1', 'Nfkb2', 'Rela', 'Rel', 'Relb']\n",
    "sc.pl.umap(adata,color=nfkb)\n",
    "sc.tl.score_genes(adata, nfkball)\n",
    "sc.pl.umap(adata, color=nfkbfactors)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata, color='Slc4a11')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "(clusters, clusterkeys, HPCS) = resetClusters('cell2020')\n",
    "scoreAndPlot(adata, numgenes=500)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "(clusters, clusterkeys, HPCS) = resetClusters('cl5bootstrap2')\n",
    "scoreAndPlot(adata,numgenes=100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata, color=['GFP'], cmap='Reds')\n",
    "sc.pl.umap(adata, color=['mKate2'], cmap='Reds')\n",
    "sc.pl.umap(adata, color=['Nfkb1'], cmap='Reds')\n",
    "sc.pl.umap(adata, color=['Nfkb1', \"2117-Nfkb1\",\"2118-Nfkb1\",'shRenilla'], cmap='Reds')\n",
    "sc.pl.umap(adata, color=[\"2118-Nfkb1\"], cmap='Reds')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "sc.pl.umap(adata, color=['n_genes', 'n_genes_by_counts', 'total_counts', 'total_counts_mt',\n",
    "       'pct_counts_mt', 'log10GenesPerUMI'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "sc.pl.umap(adata,color='Cluster 5', cmap='Reds')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata,color='Cluster 11', cmap='Reds')"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {
    "scrolled": true
   },
   "source": [
    "sc.pl.umap(adata, color='Cluster 5', cmap=\"jet\", save='Cluster 5.svg')"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "sc.pl.umap(adata, color='Slc4a11', cmap='Reds', save='Slc4a11.svg')"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "sc.pl.umap(adata, color=['Slc4a11', 'Itga2',  'Cldn4', 'Plaur', 'Krt8', 'Krt18'], cmap='Reds') #'Tigit',"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {
    "scrolled": true
   },
   "source": [
    "sc.pl.umap(adata, color=['Vim','Wnt5a','Bst2'], cmap='Reds')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.tl.embedding_density(adata, basis='umap', groupby='Classification')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.embedding_density(adata, basis='umap', key='umap_density_Classification')"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "sc.pl.umap(adata, color='GFP')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "sc.pl.matrixplot(adata, ['Nfkb1','Slc4a11','Cluster 5'], groupby='Classification', standard_scale='var')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "adata.obs.Classification.value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Clustering the neighborhood graph"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As with Seurat and many other frameworks, we recommend the Leiden graph-clustering method (community detection based on optimizing modularity) by [Traag *et al.* (2018)](https://scanpy.readthedocs.io/en/latest/references.html#traag18). Note that Leiden clustering directly clusters the neighborhood graph of cells, which we already computed in the previous section."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.tl.leiden(adata, resolution=.275)\n",
    "sc.pl.umap(adata, color=['leiden','Slc4a11','Cluster 5'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Plot the clusters, which agree quite well with the result of Seurat."
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {
    "scrolled": false
   },
   "source": [
    "sc.pl.umap(adata, color=['leiden', 'Epcam', 'Sftpc', # Tumor cells\n",
    "                         'Ptprc', #immune\n",
    "                         'Cd3e', 'Cd4', 'Cd8a', #T cell\n",
    "                         'Cd19', 'Ms4a1', 'Cd22', # B cell\n",
    "                         'Itgax', 'Il3ra', # Dendritic Cell\n",
    "                         'Ncam1', 'Ncr1', # NK Cell\n",
    "                         'Cd34', # Hematopoetic stem cell\n",
    "                         'Cd14', 'Cd33', # macrophage\n",
    "                         #'Ceacam8' # Granulocyte\n",
    "                         'Itga2b', 'Itgb3', 'Selp', #platelet\n",
    "                         #'Gypa',  # erythrocyte\n",
    "                         'Mcam', 'Vcam1', 'Pecam1', #'Sele', # endothelial cell\n",
    "                         'Cd109', 'Wnt5a', 'Kras', 'Trp53'],vmax=2,cmap=\"jet\")"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {
    "scrolled": true
   },
   "source": [
    "sc.pl.umap(adata, color=['Kras', 'Cd109', 'Zeb2', 'Sox11'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "_ = plt.hist(adata.obs['Cluster 5'], bins='auto')\n",
    "plt.ylabel(\"Count\")\n",
    "plt.title(\"Cluster 5 Score\")\n",
    "plt.axvline(x=0.2)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import libraries (some are for cosmetics)\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "from scipy import stats\n",
    "from matplotlib.ticker import (MultipleLocator, FormatStrFormatter, AutoMinorLocator)\n",
    "from scipy.stats import norm\n",
    "from sklearn.mixture import BayesianGaussianMixture\n",
    "import matplotlib as mpl\n",
    "\n",
    "x = np.array(adata.obs['Cluster 5']).reshape(-1, 1)\n",
    "\n",
    "# create GMM model object\n",
    "gmm = BayesianGaussianMixture(n_components=2, random_state=250,init_params='k-means++')\n",
    "\n",
    "# find useful parameters\n",
    "fitGMM = gmm.fit(x)\n",
    "mean = fitGMM.means_  \n",
    "covs  = fitGMM.covariances_\n",
    "weights = fitGMM.weights_\n",
    "\n",
    "# create necessary things to plot\n",
    "x_axis = np.arange(0, 1, 0.01)\n",
    "y_axis0 = norm.pdf(x_axis, float(mean[0][0]), np.sqrt(float(covs[0][0][0])))*weights[0] # 1st gaussian\n",
    "y_axis1 = norm.pdf(x_axis, float(mean[1][0]), np.sqrt(float(covs[1][0][0])))*weights[1] # 2nd gaussian\n",
    "cutoff = 0\n",
    "if mean[0][0] < mean[1][0]:\n",
    "    cutoff = x_axis[np.where(y_axis1 < y_axis0)[0][-1]+2]\n",
    "else:\n",
    "    cutoff = x_axis[np.where(y_axis0 < y_axis1)[0][-1]+2]\n",
    "    \n",
    "\n",
    "ax = fig.add_subplot(1,2,2)\n",
    "# Plot 2\n",
    "plt.hist(x, density=True, color='black', bins=100)\n",
    "plt.plot(x_axis, y_axis0, lw=3, c='C6')\n",
    "plt.plot(x_axis, y_axis1, lw=3, c='C1')\n",
    "plt.plot(x_axis, y_axis0+y_axis1, lw=3, c='C0', ls=':')\n",
    "#plt.xlim(0, 4, 0.1)\n",
    "#plt.ylim(0.0, 2.0)\n",
    "plt.xlabel(r\"Xlog10 + 1 counts\", fontsize=20)\n",
    "plt.ylabel(r\"freq\", fontsize=20)\n",
    "plt.axvline(x=cutoff)\n",
    "\n",
    "#plt.subplots_adjust(wspace=0.3)\n",
    "plt.show()\n",
    "plt.close('all')\n",
    "print(fitGMM.converged_)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs['Cluster 5 high'] = ['1' if z >cutoff else '-1' for z in adata.obs['Cluster 5']]\n",
    "\n",
    "\n",
    "sc.pl.umap(adata, color=['Cluster 5 high'], cmap=\"tab20c_r\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#fix a bug; 'base' is None but that doesn't seem to have transferred\n",
    "adata.uns['log1p'] = {}\n",
    "adata.uns['log1p']['base'] = None"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Save the result."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "sc.pl.umap(adata, color=['leiden', 'Cluster 5', 'Cluster 11'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Logistic Regression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "from sklearn.linear_model import LogisticRegression\n",
    "from joblib import dump, load\n",
    "names = load('genes.joblib')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_log = adata[:,adata.var.index.isin(names)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_names = [x for x in adata_log.var.index]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "goi = [x for x in adata_log.var.index]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load SmartSeq h5ad file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "smartseq = anndata.read('smartseq.h5ad')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Logistic Regression against Cancer Cell 2020 Clusters intersected with 10x data"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {
    "scrolled": true
   },
   "source": [
    "goi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.model_selection import train_test_split\n",
    "\n",
    "names_adata = smartseq[:,smartseq.var.index.isin(goi)]\n",
    "\n",
    "X_train, X_test, y_train, y_test = train_test_split(names_adata.X, smartseq.obs['clusterK12'], test_size=0.25, random_state=250)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "names = [x for x in names_adata.var.index]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import the class\n",
    "from sklearn.linear_model import LogisticRegression\n",
    "\n",
    "# instantiate the model (using the default parameters)\n",
    "#logreg = LogisticRegression(random_state=250)\n",
    "#logreg = LogisticRegression(random_state=250,multi_class='multinomial',solver='saga')\n",
    "logreg = LogisticRegression(random_state=250,multi_class='multinomial', solver='lbfgs')\n",
    "\n",
    "# fit the model with data\n",
    "logreg.fit(X_train, y_train)\n",
    "\n",
    "y_pred = logreg.predict(X_test)\n",
    "y_score = logreg.predict_proba(X_test)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import the metrics class\n",
    "from sklearn import metrics\n",
    "\n",
    "cnf_matrix = metrics.confusion_matrix(y_test, y_pred)\n",
    "cnf_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import required modules\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "class_names=[0,1] # name  of classes\n",
    "fig, ax = plt.subplots()\n",
    "tick_marks = np.arange(len(class_names))\n",
    "plt.xticks(tick_marks, class_names)\n",
    "plt.yticks(tick_marks, class_names)\n",
    "# create heatmap\n",
    "sns.heatmap(pd.DataFrame(cnf_matrix), annot=True, cmap=\"YlGnBu\" ,fmt='g')\n",
    "ax.xaxis.set_label_position(\"top\")\n",
    "plt.tight_layout()\n",
    "plt.title('Confusion matrix', y=1.1)\n",
    "plt.ylabel('Actual label')\n",
    "plt.xlabel('Predicted label')\n",
    "\n",
    "#Text(0.5,257.44,'Predicted label');"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.metrics import classification_report\n",
    "target_names = ['1', '2', '3','4','5','6','7','8','9','10','11','12']\n",
    "print(classification_report(y_test, y_pred, target_names=target_names))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.preprocessing import LabelBinarizer\n",
    "\n",
    "label_binarizer = LabelBinarizer().fit(y_train)\n",
    "y_onehot_test = label_binarizer.transform(y_test)\n",
    "y_onehot_test.shape  # (n_samples, n_classes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "class_of_interest = 9\n",
    "class_id = np.flatnonzero(label_binarizer.classes_ == class_of_interest)[0]\n",
    "class_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "\n",
    "from sklearn.metrics import RocCurveDisplay\n",
    "\n",
    "RocCurveDisplay.from_predictions(\n",
    "    y_onehot_test[:, class_id],\n",
    "    y_score[:, class_id],\n",
    "    name=f\"{class_of_interest} vs the rest\",\n",
    "    color=\"darkorange\",\n",
    "    #plot_chance_level=True,\n",
    ")\n",
    "plt.axis(\"square\")\n",
    "plt.xlabel(\"False Positive Rate\")\n",
    "plt.ylabel(\"True Positive Rate\")\n",
    "plt.title(\"One-vs-Rest ROC curves:\\nVirginica vs (Setosa & Versicolor)\")\n",
    "plt.legend()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Now write logit classifier to load later\n",
    "from joblib import dump, load\n",
    "dump(logreg, 'cancerCell2020Logreg_intersecting-all.joblib')\n",
    "dump(names, 'intersecting_genes_new-all.joblib')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "y_pred = logreg.predict(smartseq[:,names].X)\n",
    "y_proba = logreg.predict_proba(smartseq[:,names].X)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "smartseq.obs['clusterK12_intersecting'] = [\"%02d\" % x for x in y_pred]\n",
    "smartseq.obs['clusterK12_intersecting'] = smartseq.obs['clusterK12_intersecting'].astype('category')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#results_binary = y_pred\n",
    "#probs_binary = y_proba\n",
    "\n",
    "results_binary = logreg.predict(X_test)\n",
    "probs_binary = logreg.predict_proba(X_test)\n",
    "\n",
    "results_df_binary = pd.DataFrame()\n",
    "results_df_binary['predicted'] = list(results_binary)\n",
    "#results_df_binary['actual'] = list(smartseq.obs['clusterK12'])\n",
    "results_df_binary['actual'] = list(y_test)\n",
    "for x in range(0,12):\n",
    "    results_df_binary[\"probability_%d\" % (x+1)] = [i[x] for i in list(probs_binary)]\n",
    "results_df_binary['highest_prob'] =results_df_binary[[\"probability_%d\" % (x+1) for x in range(0,12)]].max(axis=1)\n",
    "results_df_binary['correct'] = (results_df_binary['predicted'] == results_df_binary['actual']).astype(int)\n",
    "results_df_binary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "results_df_binary['bin'] = pd.qcut(results_df_binary['highest_prob'], q=7)\n",
    "idx = pd.IntervalIndex(results_df_binary['bin'])\n",
    "results_df_binary['low'] = idx.left\n",
    "results_df_binary['high'] = idx.right\n",
    "results_df_binary['prob range'] = results_df_binary['low'].round(2).astype(str) + \"-\" + results_df_binary['high'].round(2).astype(str)\n",
    "df_bins_grouped = results_df_binary[['low','high','correct','prob range']].groupby('prob range').mean().reset_index()\n",
    "df_bins_grouped"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "plt.figure(figsize=(4,4))\n",
    "plt.subplots_adjust(top=0.85)\n",
    "#plt.figure(figsize=(11,9))\n",
    "#plt.subplots_adjust(top=0.85)\n",
    "\n",
    "ax = sns.barplot(x='prob range', y='correct', data=df_bins_grouped)\n",
    "plt.title(\"Logistic Regression Accuracy by Probability Range\")\n",
    "plt.ylabel('Percent correct')\n",
    "plt.xlabel('Probability range')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#reset size\n",
    "from matplotlib import rcParams\n",
    "FIGSIZE=(4,4)\n",
    "rcParams['figure.figsize']=FIGSIZE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(smartseq, color=['clusterK12','clusterK12_intersecting'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dict = {1: \"01\", 2: \"02\", 3: \"03\", 4: \"04\", 5: \"05\", 6: \"06\", 7: \"07\", 8: \"08\", 9: \"09\", 10: \"10\", 11: \"11\", 12: \"12\"}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.metrics import accuracy_score\n",
    "accuracy_score(smartseq.obs.replace({\"clusterK12\": dict})['clusterK12'], smartseq.obs['clusterK12_intersecting'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "results_binary = logreg.predict(smartseq[:,names].X)\n",
    "probs_binary = logreg.predict_proba(smartseq[:,names].X)\n",
    "\n",
    "results_df_binary = pd.DataFrame()\n",
    "results_df_binary['predicted'] = list(results_binary)\n",
    "#results_df_binary['actual'] = list(smartseq.obs['clusterK12'])\n",
    "for x in range(0,12):\n",
    "    results_df_binary[x+1] = [i[x] for i in list(probs_binary)]\n",
    "results_df_binary['highest_prob'] =results_df_binary[[x+1 for x in range(0,12)]].max(axis=1)\n",
    "results_df_binary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "_ = plt.hist(results_df_binary['highest_prob'], bins='auto', histtype='step')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## End LogisticRegression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_ordered_genes = [adata_names.index(x) for x in names]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_log_ordered = adata_log[:,adata_ordered_genes]"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {
    "scrolled": true
   },
   "source": [
    "adata_log_ordered.var.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "y_pred = logreg.predict(adata_log_ordered.X)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "adata_log_ordered.obs['clusterK12'] = [\"%02d\" % x for x in y_pred]\n",
    "adata_log_ordered.obs['clusterK12'] = adata_log_ordered.obs['clusterK12'].astype('category')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#reset size\n",
    "from matplotlib import rcParams\n",
    "FIGSIZE=(4,4)\n",
    "rcParams['figure.figsize']=FIGSIZE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata_log_ordered,color=['leiden','clusterK12'], wspace=0.1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "sc.pl.umap(adata_log_ordered,color=['clusterK12'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for x in adata_log_ordered.obs.clusterK12.cat.categories:\n",
    "    sc.pl.umap(adata_log_ordered,color=['clusterK12'], groups=[x])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# bug fix\n",
    "adata_log_ordered.uns['log1p']['base'] = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    " cell_proportion_df = adata_log_ordered.obs.leiden.value_counts().T.plot(kind='bar', stacked=True, legend=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    " cell_proportion_df = adata_log_ordered.obs.clusterK12.value_counts(sort=False).T.plot(kind='bar', stacked=True, legend=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    " cell_proportion_df = pd.crosstab(adata_log_ordered.obs['leiden'],adata_log_ordered.obs['clusterK12'], normalize='columns').T.plot(kind='bar', stacked=True).legend(bbox_to_anchor=(1,1))"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "Now consider there may be some predictions that are poor"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "results_binary = logreg.predict(adata_log_ordered.X)\n",
    "probs_binary = logreg.predict_proba(adata_log_ordered.X)\n",
    "\n",
    "results_df_binary = pd.DataFrame()\n",
    "results_df_binary['predicted'] = list(results_binary)\n",
    "#results_df_binary['actual'] = list(smartseq.obs['clusterK12'])\n",
    "for x in range(0,12):\n",
    "    results_df_binary[x+1] = [i[x] for i in list(probs_binary)]\n",
    "results_df_binary['highest_prob'] =results_df_binary[[x+1 for x in range(0,12)]].max(axis=1)\n",
    "results_df_binary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "_ = plt.hist(results_df_binary['highest_prob'], bins='auto')\n",
    "plt.axvline(x=0.9)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "results_df_binary['bin'] = pd.qcut(results_df_binary['highest_prob'], q=7)\n",
    "idx = pd.IntervalIndex(results_df_binary['bin'])\n",
    "results_df_binary['low'] = idx.left\n",
    "results_df_binary['high'] = idx.right\n",
    "results_df_binary['prob range'] = results_df_binary['low'].round(2).astype(str) + \"-\" + results_df_binary['high'].round(2).astype(str)\n",
    "df_bins_grouped = results_df_binary[['low','high','prob range']].groupby('prob range').mean().reset_index()\n",
    "df_bins_grouped"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cutoff = 0.8\n",
    "adata_log_ordered.obs['clusterK12_stringent'] = 'other'\n",
    "adata_log_ordered.obs['clusterK12_stringent'][np.array(results_df_binary['highest_prob'] >= cutoff)] = [\"%02d\" % y for y in results_df_binary[[x+1 for x in range(0,12)]].idxmax(axis=1)[np.array(results_df_binary['highest_prob'] >= cutoff)].to_numpy()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_log_ordered.obs['clusterK12_stringent']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_log_ordered.obs['clusterK12_stringent'] = adata_log_ordered.obs['clusterK12_stringent'].astype('category')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata_log_ordered, color=['clusterK12', 'clusterK12_stringent'], wspace=0.5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    " cell_proportion_df = pd.crosstab(adata_log_ordered.obs['clusterK12_stringent'],adata_log_ordered.obs['clusterK12'],normalize='columns').T.plot(kind='bar', stacked=True).legend(bbox_to_anchor=(1,1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    " cell_proportion_df = pd.crosstab(adata_log_ordered.obs['leiden'],adata_log_ordered.obs['clusterK12_stringent'], normalize='columns').T.plot(kind='bar', stacked=True).legend(bbox_to_anchor=(1,1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    " cell_proportion_df = pd.crosstab(adata_log_ordered.obs['Group'],adata_log_ordered.obs['clusterK12_stringent'], normalize='columns').T.plot(kind='bar', stacked=True).legend(bbox_to_anchor=(1,1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cancerCell = ['Cluster 1', 'Cluster 2','Cluster 3','Cluster 4','Cluster 5','Cluster 6', 'Cluster 7', \n",
    "              'Cluster 8', 'Cluster 9', 'Cluster 10', 'Cluster 11', 'Cluster 12']\n",
    "#cell_proportion_df = pd.crosstab(adata_log_ordered.obs['Cluster 1'],adata_log_ordered.obs['clusterK12']).T.plot(kind='bar', stacked=True).legend(bbox_to_anchor=(1,1))\n",
    "sc.pl.matrixplot(adata_log_ordered, cancerCell, 'clusterK12', standard_scale='var')\n",
    "sc.pl.matrixplot(adata_log_ordered, cancerCell, 'clusterK12_stringent', standard_scale='var')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for x in adata_log_ordered.obs.clusterK12_stringent.cat.categories:\n",
    "    # skip if NA\n",
    "    sc.pl.umap(adata_log_ordered, color='clusterK12_stringent', groups=[x])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.tl.rank_genes_groups(adata_log_ordered, 'clusterK12', method='logreg')\n",
    "sc.pl.rank_genes_groups(adata_log_ordered, n_genes=25, sharey=False)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {
    "scrolled": true
   },
   "source": [
    "sc.tl.rank_genes_groups(adata_log_ordered, 'clusterK12_stringent', method='logreg')\n",
    "sc.pl.rank_genes_groups(adata_log_ordered, n_genes=25, sharey=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "adata_log_ordered.obs.clusterK12.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "cancerCell = ['Cluster 1', 'Cluster 2','Cluster 3','Cluster 4','Cluster 5','Cluster 6', 'Cluster 7', \n",
    "              'Cluster 8', 'Cluster 9', 'Cluster 10', 'Cluster 11', 'Cluster 12']\n",
    "#sc.pl.violin(adata_log_ordered, cancerCell, groupby='clusterK12')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.stacked_violin(adata_log_ordered, cancerCell, groupby='clusterK12_stringent')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "sc.pl.dotplot(adata_log_ordered, cancerCell, groupby='clusterK12_stringent')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.matrixplot(adata_log_ordered, cancerCell, groupby='Group',standard_scale='var')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.matrixplot(adata_log_ordered, ['Nfkb1','Nfkbia'], groupby='Group',standard_scale='var')"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "import matplotlib.pyplot as plt\n",
    "_ = plt.hist(adata_log_ordered.obs['Cluster 2'], bins='auto')\n",
    "plt.ylabel(\"Count\")\n",
    "plt.title(\"Cluster 2 Score\")\n",
    "plt.axvline(x=0.15)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "from sklearn.mixture import BayesianGaussianMixture\n",
    "X = np.array(adata_log_ordered.obs['Cluster 5']).reshape(-1,1)\n",
    "bgm = BayesianGaussianMixture(n_components=2, random_state=250).fit(X)\n",
    "#bgm.predict(X)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "adata.obs['clusters'].value_counts()"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "for x in adata.obs['clusters'].value_counts().index:\n",
    "    sc.pl.umap(adata, color='clusters', groups=[x])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs['clusterK12'] = adata_log_ordered.obs['clusterK12'].astype('category')\n",
    "adata.obs['clusterK12_stringent'] = adata_log_ordered.obs['clusterK12_stringent'].astype('category')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    " cell_proportion_df = pd.crosstab(adata.obs['leiden'],adata_log_ordered.obs['clusterK12_stringent'], normalize='columns').T.plot(kind='bar', stacked=True).legend(bbox_to_anchor=(1,1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    " cell_proportion_df = pd.crosstab(adata.obs['Group'],adata_log_ordered.obs['clusterK12_stringent'], normalize='columns').T.plot(kind='bar', stacked=True).legend(bbox_to_anchor=(1,1))"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "sc.pl.dotplot(adata, ['Slc4a11','Itga2','Cldn4'],groupby=['clusterK12'])"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "sc.pl.dotplot(adata, ['Slc4a11','Itga2','Cldn4'],groupby=['clusterK12_stringent'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Now retrain a new Logit classifier based off the single cell 10x data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Logistic Regression against Cancer Cell 2020 Clusters in 10x data (stringent)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "adata.obs.Group.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.model_selection import train_test_split\n",
    "\n",
    "names_adata = adata[:,~adata.var.index.isin(['mKate2','Cre','GFP'])]\n",
    "names_adata = names_adata[names_adata.obs.Group.isin(['Renilla_Control']),:]\n",
    "\n",
    "X_train, X_test, y_train, y_test = train_test_split(names_adata[~(names_adata.obs['clusterK12_stringent'] == 'other'),:].X, names_adata[~(names_adata.obs['clusterK12_stringent'] == 'other'),:].obs['clusterK12'], test_size=0.25, random_state=250)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "names = [x for x in names_adata.var.index]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# import the class\n",
    "from sklearn.linear_model import LogisticRegression\n",
    "\n",
    "# instantiate the model (using the default parameters)\n",
    "logreg = LogisticRegression(random_state=250, multi_class='multinomial', solver='saga')\n",
    "\n",
    "# fit the model with data\n",
    "logreg.fit(X_train, y_train)\n",
    "\n",
    "y_pred = logreg.predict(X_test)\n",
    "y_score = logreg.predict_proba(X_test)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.DataFrame(y_pred).value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import the metrics class\n",
    "from sklearn import metrics\n",
    "\n",
    "cnf_matrix = metrics.confusion_matrix(y_test, y_pred)\n",
    "cnf_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import required modules\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "class_names=[0,1] # name  of classes\n",
    "fig, ax = plt.subplots()\n",
    "tick_marks = np.arange(len(class_names))\n",
    "plt.xticks(tick_marks, class_names)\n",
    "plt.yticks(tick_marks, class_names)\n",
    "# create heatmap\n",
    "sns.heatmap(pd.DataFrame(cnf_matrix), annot=True, cmap=\"YlGnBu\" ,fmt='g')\n",
    "ax.xaxis.set_label_position(\"top\")\n",
    "plt.tight_layout()\n",
    "plt.title('Confusion matrix', y=1.1)\n",
    "plt.ylabel('Actual label')\n",
    "plt.xlabel('Predicted label')\n",
    "\n",
    "#Text(0.5,257.44,'Predicted label');"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {
    "scrolled": false
   },
   "source": [
    "from sklearn.metrics import classification_report\n",
    "target_names = ['1', '2', '3','4','5','6','7','8','9','10','11','12']\n",
    "print(classification_report(y_test, y_pred, labels=target_names))"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "y_pred_proba = logreg.predict_proba(X_test)[::,1]\n",
    "fpr, tpr, _ = metrics.roc_curve(y_test,  y_pred_proba)\n",
    "auc = metrics.roc_auc_score(y_test, y_pred_proba)\n",
    "plt.plot(fpr,tpr,label=\"data 1, auc=\"+str(auc))\n",
    "plt.legend(loc=4)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.preprocessing import LabelBinarizer\n",
    "\n",
    "label_binarizer = LabelBinarizer().fit(y_train)\n",
    "y_onehot_test = label_binarizer.transform(y_test)\n",
    "y_onehot_test.shape  # (n_samples, n_classes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "label_binarizer.classes_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "class_of_interest = '05'\n",
    "class_id = np.flatnonzero(label_binarizer.classes_ == class_of_interest)[0]\n",
    "class_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "\n",
    "from sklearn.metrics import RocCurveDisplay\n",
    "\n",
    "RocCurveDisplay.from_predictions(\n",
    "    y_onehot_test[:, class_id],\n",
    "    y_score[:, class_id],\n",
    "    name=f\"{class_of_interest} vs the rest\",\n",
    "    color=\"darkorange\",\n",
    "    #plot_chance_level=True,\n",
    ")\n",
    "plt.axis(\"square\")\n",
    "plt.xlabel(\"False Positive Rate\")\n",
    "plt.ylabel(\"True Positive Rate\")\n",
    "plt.title(\"One-vs-Rest ROC curves:\\nVirginica vs (Setosa & Versicolor)\")\n",
    "plt.legend()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Now write logit classifier to load later\n",
    "from joblib import dump, load\n",
    "dump(logreg, 'cancerCell2020Logreg_retrain-all.joblib')\n",
    "dump(names, 'genes_retrain-all.joblib')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "y_pred = logreg.predict(adata.X)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs['clusterK12_retrain'] = y_pred"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs['clusterK12_retrain'] = adata.obs['clusterK12_retrain'].astype('category')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata, color = ['clusterK12','clusterK12_retrain'], wspace=0.5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    " cell_proportion_df = adata.obs.clusterK12.value_counts(sort=False).T.plot(kind='bar', stacked=True, legend=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    " cell_proportion_df = adata.obs.clusterK12_retrain.value_counts(sort=False).T.plot(kind='bar', stacked=True, legend=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    " cell_proportion_df = pd.crosstab(adata.obs['clusterK12_retrain'],adata.obs['clusterK12'], normalize='columns').T.plot(kind='bar', stacked=True).legend(bbox_to_anchor=(1,1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.stacked_violin(adata, cancerCell, 'clusterK12')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.stacked_violin(adata, cancerCell, 'clusterK12_retrain')"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {
    "scrolled": false
   },
   "source": [
    "sc.tl.rank_genes_groups(adata, 'clusterK12', method='wilcoxon')\n",
    "sc.pl.rank_genes_groups(adata, n_genes=25, sharey=False)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {
    "scrolled": false
   },
   "source": [
    "sc.tl.rank_genes_groups(adata, 'clusterK12_retrain', method='wilcoxon')\n",
    "sc.pl.rank_genes_groups(adata, n_genes=25, sharey=False)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "pd.DataFrame(bgm.predict(X)).value_counts()"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "_ = plt.hist(bgm.predict(X), bins='auto')\n",
    "plt.ylabel(\"Count\")\n",
    "plt.title(\"BGM Prediction\")\n",
    "plt.axvline(x=0.15)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## End Logistic Regression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata, color=['Onecut2'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata, color=['Slc4a11','Cldn4','Itga2', 'Plaur']) #'Tigit',"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "sc.pl.umap(adata, color=['Krt8','Krt18','Krt7','Krt19', 'Krt20', 'Krt23', 'Krt14', 'Nkx2-1'], cmap=\"jet\") #'Krt5', "
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "sc.pl.umap(adata, color=['Anxa3', 'Anxa11', 'Cd63','Cd9', 'Anxa1','Vim','S100a6','S100a11', 'Casp3', 'Casp9'], cmap=\"jet\")"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "sc.pl.umap(adata, color=['Anxa2', 'Anxa1', 'Anxa5','Anxa3', 'Anxa7'], cmap=\"jet\")"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "sc.pl.umap(adata, color=['S100a6','S100a14','S100a10','S100a11','S100a16'], cmap=\"jet\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "sc.pl.umap(adata, color=['Plaur','Itgb1','Cd151', 'Cd9'], cmap=\"jet\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata, color=['Vim',], cmap=\"jet\") #'Cd109'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "sc.pl.umap(adata, color=['Hopx','Pdpn'], cmap=\"jet\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata, color=['Smad2', 'Smad3'], cmap=\"jet\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.write_h5ad(results_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for label in adata.obs['Classification'].unique():\n",
    "    sc.pl.umap(adata, color='Classification',groups=[label])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "sc.tl.embedding_density(adata,basis='umap', groupby='Group')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs.Group.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.embedding_density(adata,basis='umap',key='umap_density_Group', \n",
    "                        group=['Renilla_Control','2117_Control','2118_Control',\n",
    "                              'Renilla_EGFRi','2117_EGFRi','2118_EGFRi'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "sc.tl.embedding_density(adata,basis='umap', groupby='clusterK12')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.embedding_density(adata,basis='umap',key='umap_density_clusterK12')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs[['Group', 'clusterK12']].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "sc.pl.violin(adata, ['Nfkb1','Rel','Rela','Nfkb2','Relb','Cluster 5','Cl5Bootstrap2'], groupby='Group',rotation=90,inner='box',stripplot=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.dotplot(adata, ['Nfkb1','Rel','Rela','Nfkb2','Relb','Nfkbia','Tnf','Il1r1'], groupby='Group') #'Il1b'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.violin(adata, ['Nfkbia'],groupby='Group', rotation=90, stripplot=False, inner='box')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "sc.pl.matrixplot(adata, ['Nfkb1','Nfkbia','Slc4a11','Itga2','Cluster 5','Cl5Bootstrap2'], groupby='Group', standard_scale='var')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.dotplot(adata, ['Nfkb1','Nfkbia','Slc4a11','Itga2','Cluster 5'], groupby='Group', standard_scale='var')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "sc.pl.dotplot(adata, ['Nfkb1','Nfkbia','Slc4a11','Itga2','Cluster 5'], groupby='Group', standard_scale='var')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs.Group.value_counts()"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "import pickle\n",
    "with open('write/cellwhitelist-lane1.pickle', 'wb') as f:\n",
    "    # Pickle the 'data' dictionary using the highest protocol available.\n",
    "    pickle.dump(tuple(adata.obs.index), f, pickle.HIGHEST_PROTOCOL)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "internalhash = None\n",
    "with open('CellRanger output/hashIDs.pickle', 'rb') as f:\n",
    "    internalhash = pickle.load(f)\n",
    "internalhash"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "nadata = adata[internalhash.index,:]"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "classify = np.load('CellRanger output/classification.npy', allow_pickle=True)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "classify"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "nadata.obs[\"shRNA\"] = classify\n",
    "nadata.obs[\"shRNA\"].value_counts()"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {
    "scrolled": true
   },
   "source": [
    "sc.pl.umap(nadata, color=['Group', 'shRNA'])"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "sc.pl.matrixplot(nadata, ['Nfkb1','Slc4a11','Cluster 5','Cluster 11', 'mKate2','GFP'], groupby='Group', standard_scale='var')"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "nadata.obs.Group.value_counts()"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {
    "scrolled": true
   },
   "source": [
    "sc.pl.umap(nadata,color='GFP')"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {
    "scrolled": true
   },
   "source": [
    "sc.pl.umap(adata,color='GFP')"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "sc.pl.umap(adata, color=['Cre','GFP'])\n",
    "#sc.pl.scatter(adata, x='Cre', y='GFP')"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "import matplotlib.pyplot as plt\n",
    "#Remove any cell with PTPRC counts\n",
    "adata.var.index.isin(['GFP'])\n",
    "_ = plt.hist(adata.raw.to_adata()[adata.obs.index,'GFP'].X.toarray())\n",
    "plt.ylabel(\"Count\")\n",
    "plt.title(\"GFP\")\n",
    "plt.axvline(x=1)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {
    "scrolled": true
   },
   "source": [
    "import matplotlib.pyplot as plt\n",
    "#Remove any cell with PTPRC counts\n",
    "adata.var.index.isin(['Cre'])\n",
    "_ = plt.hist(adata.raw.to_adata()[adata.obs.index,'Cre'].X.toarray())\n",
    "plt.ylabel(\"Count\")\n",
    "plt.title(\"Cre\")\n",
    "plt.axvline(x=1)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "#PtprcCells = adata[adata.X[:,adata.var.index.isin(['Ptprc'])] >= 0.5,:].obs.index\n",
    "CreCells = adata.raw.to_adata()[adata.raw.to_adata()[adata.obs.index,'Cre'].X >= 0.5,:].obs.index"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "nadata2 = adata[CreCells,:]\n",
    "nadata2"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "sc.pl.umap(nadata2, color=['Cre', 'GFP'])"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {
    "scrolled": true
   },
   "source": [
    "sc.pl.dotplot(nadata2, ['Nfkb1','Slc4a11','Cluster 5','Cluster 11', 'mKate2','GFP'], groupby='Group')"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {
    "scrolled": true
   },
   "source": [
    "sc.pl.violin(nadata2, ['GFP', 'Cre', 'Cluster 5', 'Cluster 11', 'Nfkb1'], groupby='Group', inner='box',stripplot=False, rotation=90)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "sc.pl.violin(nadata2, ['GFP', 'Cre', 'Cl5Bootstrap2', 'Cluster 11', 'Nfkb1'], groupby='Group',stripplot=False, inner='box', rotation=90)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "nadata2.obs.Group.value_counts()"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "nadata.obs.Group.value_counts()"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "sc.pl.umap(nadata,color='GFP')"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {
    "scrolled": true
   },
   "source": [
    "# the next 3 cells break h5ad writing capability\n",
    "sc.tl.rank_genes_groups(adata, 'Cluster 5 high', method='wilcoxon')"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "sc.tl.filter_rank_genes_groups(adata, groupby='Cluster 5 high', max_out_group_fraction=0.2)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "result = adata.uns['rank_genes_groups_filtered']\n",
    "groups = result['names'].dtype.names\n",
    "pd.DataFrame(\n",
    "    {group + '_' + key[:1]: result[key][group]\n",
    "    for group in groups for key in ['names', 'pvals']}).head(200).to_excel('test.xlsx')"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "## Finding marker genes"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "Let us compute a ranking for the highly differential genes in each cluster. For this, by default, the `.raw` attribute of AnnData is used in case it has been initialized before. The simplest and fastest method to do so is the t-test."
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "# bug fix\n",
    "adata.uns['log1p']['base'] = None"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {
    "scrolled": true
   },
   "source": [
    "sc.tl.rank_genes_groups(adata, 'leiden', method='t-test')\n",
    "sc.pl.rank_genes_groups(adata, n_genes=25, sharey=False)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "sc.settings.verbosity = 2  # reduce the verbosity"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "The result of a [Wilcoxon rank-sum (Mann-Whitney-U)](https://en.wikipedia.org/wiki/Mann%E2%80%93Whitney_U_test) test is very similar. We recommend using the latter in publications, see e.g., [Sonison & Robinson (2018)](https://doi.org/10.1038/nmeth.4612). You might also consider much more powerful differential testing packages like MAST, limma, DESeq2 and, for python, the recent diffxpy."
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "sc.tl.rank_genes_groups(adata, 'leiden', method='wilcoxon')\n",
    "sc.pl.rank_genes_groups(adata, n_genes=25, sharey=False)\n",
    "sc.pl.rank_genes_groups_dotplot(adata, n_genes=5, groupby=\"leiden\")\n"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "Save the result."
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "adata.write(results_file, compression='gzip')"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "As an alternative, let us rank genes using logistic regression. For instance, this has been suggested by [Natranos et al. (2018)](https://doi.org/10.1101/258566). The essential difference is that here, we use a multi-variate appraoch whereas conventional differential tests are uni-variate. [Clark et al. (2014)](https://doi.org/10.1186/1471-2105-15-79) has more details."
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "sc.tl.rank_genes_groups(adata, 'leiden', method='logreg')\n",
    "sc.pl.rank_genes_groups(adata, n_genes=25, sharey=False)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "Reload the object that has been save with the Wilcoxon Rank-Sum test result."
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "adata = sc.read(results_file)\n",
    "\n",
    "#Export for SCENIC analysis\n",
    "adata.write_h5ad(\"write/postSCT-tumor-leiden.h5ad\",compression='gzip')\n",
    "adata = sc.read(\"write/postSCT-tumor-leiden.h5ad\")"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "nfkbgenes = ['Nfkbia', 'Bcl2', 'Bcl2l1','Xiap', ]\n",
    "inflammatory = ['Tnf', 'Il6', 'Cxcl1', 'Il6']\n",
    "\n",
    "sc.pl.violin(adata, nfkbgenes, groupby='Group', inner='box',stripplot=False,rotation=90)\n",
    "sc.pl.violin(adata, inflammatory, groupby='Group', inner='box',stripplot=False, rotation=90)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
